src/arch/riscv/sbi.c

```diff
@@ -405,6 +405,7 @@ static struct sbiret sbi_bao_handler(unsigned long fid)
     struct sbiret ret;
 
     ret.error = hypercall(fid);
+    ret.value = (long int)vcpu_readreg(cpu()->vcpu, REG_A1);
```

Comment:
I don't get your point. I don't see any other ecall handler function that does that.

Reply:
I was referring to the `sbi_vs_handler` function, which already uses `ret.value` to write to the `REG_A1` register in the final line before the return statement. Instead of removing this line, I read the value into the `sbiret` ret struct first to prevent it from being overwritten. This is necessary because writing to the register in the `remio_handle_ask` function would overwrite the same value if `vcpu_writereg(cpu()->vcpu, REG_A1, (unsigned long)ret.value)` remains.

Reply:
> I see your point. It seems we are actually breaking the [SBI spec](https://github.com/riscv-non-isa/riscv-sbi-doc/blob/master/src/binary-encoding.adoc), as it states "All registers except a0 & a1 must be preserved across an SBI call by the callee.". There is no way around this as the alternative would be using a pointer to some shared memory area which is something we really don't want. We should make sure a1 contains a value part of Bao's hypercall error codes (return status as you mention). Then the remaining return values are returned following the logic of the normal RISC-V ABI (a2, a3, etc...). So maybe we should change this macro to `#define HYPCALL_OUT_ARG_REG(ARG) (HYPCALL_IN_ARG_REG((ARG) + 2))`. What do you think?

Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/47f09052fc338b4e458bb60f3f66bf7c20ad5b25